pkg/helpers/struct.go

```diff
@@ -208,49 +207,80 @@ func AsSlice(x *messages.ListValue) []interface{} {
 // When converting an int64 or uint64 to a NumberValue, numeric precision loss
 // is possible since they are stored as a float64.
 func NewValue(v interface{}) (*messages.Value, error) {
-	switch v := v.(type) {
-	case nil:
+
+	if v == nil {
 		return NewNullValue(), nil
-	case bool:
-		return NewBoolValue(v), nil
-	case int:
-		return NewNumberValue(float64(v)), nil
-	case int32:
-		return NewNumberValue(float64(v)), nil
-	case int64:
-		return NewNumberValue(float64(v)), nil
-	case uint:
-		return NewNumberValue(float64(v)), nil
-	case uint32:
-		return NewNumberValue(float64(v)), nil
-	case uint64:
-		return NewNumberValue(float64(v)), nil
-	case float32:
-		return NewNumberValue(float64(v)), nil
-	case float64:
-		return NewNumberValue(float64(v)), nil
-	case string:
-		if !utf8.ValidString(v) {
+	}
+
+	switch reflect.TypeOf(v).Kind() {
+	case reflect.Bool:
+		return NewBoolValue(v.(bool)), nil
+	case reflect.Int:
+		return NewNumberValue(float64(v.(int))), nil
+	case reflect.Int32:
+		return NewNumberValue(float64(v.(int32))), nil
+	case reflect.Int64:
+		return NewNumberValue(float64(v.(int64))), nil
+	case reflect.Uint:
+		return NewNumberValue(float64(v.(uint))), nil
+	case reflect.Uint32:
+		return NewNumberValue(float64(v.(uint32))), nil
+	case reflect.Uint64:
+		return NewNumberValue(float64(v.(uint64))), nil
+	case reflect.Float32:
+		return NewNumberValue(float64(v.(float32))), nil
+	case reflect.Float64:
+		return NewNumberValue(v.(float64)), nil
+	case reflect.String:
+		refStr := v.(string)
+		if !utf8.ValidString(refStr) {
 			return nil, protoimpl.X.NewError("invalid UTF-8 in string: %q", v)
 		}
-		return NewStringValue(v), nil
-	case time.Time:
-		return NewTimestampValue(v), nil
-	case []byte:
-		s := base64.StdEncoding.EncodeToString(v)
-		return NewStringValue(s), nil
-	case map[string]interface{}:
-		v2, err := NewStruct(v)
-		if err != nil {
-			return nil, err
+		return NewStringValue(refStr), nil
+	case reflect.Struct:
+		// special case for time values
+		if tVal, ok := v.(time.Time); ok {
+			return NewTimestampValue(tVal), nil
 		}
-		return NewStructValue(v2), nil
-	case []interface{}:
-		v2, err := NewList(v)
+		// should we try to handle struct values somehow?
+		return nil, protoimpl.X.NewError("invalid type: %T", v)
+
+	case reflect.Map:
```

Comment:
Do we need these changes or are we just using this wrong? This function is a clone of the implementation in https://pkg.go.dev/google.golang.org/protobuf/types/known/structpb#NewValue with a single special case added for values of type time.Time (above). 

See the source here https://github.com/protocolbuffers/protobuf-go/blob/v1.28.1/types/known/structpb/struct.pb.go#L316

We are more or less making even more modifications to the protobuf standard library, which is fine if it's actually necessary, but I'm not sure I understand why we need to change this yet.

Do any of the other helpers that already exist in https://pkg.go.dev/google.golang.org/protobuf/types/known/structpb solve the problem? 

The description mentions `[]string`, so perhaps https://pkg.go.dev/google.golang.org/protobuf/types/known/structpb#ListValue?

Reply:
>  This function is a clone of the implementation in https://pkg.go.dev/google.golang.org/protobuf/types/known/structpb#NewValue with a single special case added for values of type time.Time (above).

Huh, didn't notice that.

Yah, it seems like there should be an easier way to do this. It looks like we also have our own version of `Value` defined in the protobuf files here as `messages.Value`, I'm not sure why we're not using the `Value` type defined in the protobuf libraries. I imagine that if we were, it would be a tad easier to lean on some of the other libraries.

Reply:
We defined our own value to add the Timestamp case:

```go
case time.Time:
		return NewTimestampValue(v), nil
```

This will serialize `time.Time` values as a pair of integers using the built in protobuf Timestamp type. Without that clients would need to manually construct an efficient format for timestamps, or worse just encode it to string and then parse the time string back out into a time.Time value for every event.

Reply:
Ahhh, okay, that makes sense.

Reply:
So, it looks like most of the code in `struct.go` is just the same as the structpb code, but with the changed `Value` type. So, it looks like the underlying "upstream" `NewValue` has the same issue, namely that it'll fail on `[]string` types. Looking at the example code in the structpb library, it expects a lot of stuff to get passed as `interface{}` types.

I share your concern though; I'm a tad worried we'll outgrow this particular implementation as we get weirder nested types, other libraries using it in unexpected ways, etc.

Reply:
https://github.com/golang/protobuf/issues/1302 discusses this.

We could extend this but I am now worried we'll hit some type we did not expect and get this error at runtime down the road. structpb is supposed to be the protobuf representation of a JSON object, so the recommend way to handle arbitrary types is to actually marshal them to JSON and then call `structpb.Value.UnmarshalJSON`. I think that should be our fallback path to ensure we can always encode we would not want to hit it regularly.

I wonder if we are better off converting `messages.Struct fields;` to `bytes fields_json` and passing marshalled JSON over the wire. The use of `google.protobuf.Struct` is supposed to give us a path to eliminating the use of the `mapstr` type by having an actually limited set of types to deal with, so I'd like to keep it but not at the cost of randomly being unable to serialize events or having to marshal to JSON and then unmarshal and then serialize the protobuf regularly.

Reply:
Yah, none of the solutions here feel particularly elegant; at least if we do everything as `bytes fields_json` we're not dealing with a bunch of conditional logic to step around shortcomings in various libraries. However, passing around blobs of unstructured data could be a pain for debugging, and it just doesn't feel great, and also involves some changes to the underlying protobuf API. However, it also seems like it might be the "safest" solution. 

Considering how early in development the shipper is, we could just wait and see how much pain the custom `messages.Struct` type causes us, but we'll also be more reluctant to make breaking changes to the client protobuf as time goes on.

Reply:
Yeah I am fine with running with this and seeing how we feel. A lot of this will depend on performance as well, which we are just starting to properly measure.

I think one concrete improvement we can make to this is to change the `default` case here to fall back to marshalling to JSON and then calling the [struct's UnmarshalJSON](https://pkg.go.dev/google.golang.org/protobuf/types/known/structpb#Struct.UnmarshalJSON) method to guarantee we can convert. I'd rather be slow than lose data.